From fce11f68491b46b93df69de0630cd9edb90bc772 Mon Sep 17 00:00:00 2001
From: Birger Koblitz <git@birger-koblitz.de>
Date: Wed, 29 Dec 2021 21:54:21 +0100
Subject: [PATCH] realtek: Create 4 different Realtek Platforms

Creates RTL83XX as a basic kernel config parameter for the
RTL838X, RTL839x, RTL930X and RTL931X platforms with respective
configurations for the SoCs, which are introduced in addition.

Submitted-by: Birger Koblitz <git@birger-koblitz.de>
---
--- a/arch/mips/Kbuild.platforms
+++ b/arch/mips/Kbuild.platforms
@@ -22,6 +22,7 @@ platform-$(CONFIG_MACH_NINTENDO64)	+= n6
 platform-$(CONFIG_PIC32MZDA)		+= pic32/
 platform-$(CONFIG_RALINK)		+= ralink/
 platform-$(CONFIG_MIKROTIK_RB532)	+= rb532/
+platform-$(CONFIG_MACH_REALTEK_RTL)	+= rtl838x/
 platform-$(CONFIG_SGI_IP22)		+= sgi-ip22/
 platform-$(CONFIG_SGI_IP27)		+= sgi-ip27/
 platform-$(CONFIG_SGI_IP28)		+= sgi-ip22/
--- a/arch/mips/Kconfig
+++ b/arch/mips/Kconfig
@@ -620,21 +620,23 @@ config RALINK
 
 config MACH_REALTEK_RTL
 	bool "Realtek RTL838x/RTL839x based machines"
-	select MIPS_GENERIC
 	select DMA_NONCOHERENT
 	select IRQ_MIPS_CPU
-	select CSRC_R4K
-	select CEVT_R4K
 	select SYS_HAS_CPU_MIPS32_R1
 	select SYS_HAS_CPU_MIPS32_R2
 	select SYS_SUPPORTS_BIG_ENDIAN
 	select SYS_SUPPORTS_32BIT_KERNEL
 	select SYS_SUPPORTS_MIPS16
-	select SYS_SUPPORTS_MULTITHREADING
-	select SYS_SUPPORTS_VPE_LOADER
 	select BOOT_RAW
 	select PINCTRL
 	select USE_OF
+	select NO_EXCEPT_FILL
+	select SYS_SUPPORTS_HIGHMEM
+	select SYS_HAS_EARLY_PRINTK
+	select SYS_HAS_EARLY_PRINTK_8250
+	select USE_GENERIC_EARLY_PRINTK_8250
+	select ARCH_HAS_RESET_CONTROLLER
+	select RESET_CONTROLLER
 
 config SGI_IP22
 	bool "SGI IP22 (Indy/Indigo2)"
@@ -970,6 +972,36 @@ config CAVIUM_OCTEON_SOC
 
 endchoice
 
+config RTL838X
+	bool "Realtek RTL838X based platforms"
+	depends on MACH_REALTEK_RTL
+	select CPU_SUPPORTS_CPUFREQ
+	select MIPS_EXTERNAL_TIMER
+
+config RTL839X
+	bool "Realtek RTL839X based platforms"
+	depends on MACH_REALTEK_RTL
+	select CPU_SUPPORTS_CPUFREQ
+	select MIPS_EXTERNAL_TIMER
+	select SYS_SUPPORTS_MULTITHREADING
+
+config RTL930X
+	bool "Realtek RTL930X based platforms"
+	depends on MACH_REALTEK_RTL
+	select MIPS_CPU_SCACHE
+	select MIPS_EXTERNAL_TIMER
+	select SYS_SUPPORTS_MULTITHREADING
+
+config RTL931X
+	bool "Realtek RTL931X based platforms"
+	depends on RTL930X
+	select MIPS_GIC
+	select COMMON_CLK
+	select CLKSRC_MIPS_GIC
+	select SYS_SUPPORTS_VPE_LOADER
+	select SYS_SUPPORTS_SMP
+	select SYS_SUPPORTS_MIPS_CPS
+
 source "arch/mips/alchemy/Kconfig"
 source "arch/mips/ath25/Kconfig"
 source "arch/mips/ath79/Kconfig"
